runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ControlClientPool.java

```diff
@@ -17,16 +17,58 @@
  */
 package org.apache.beam.runners.fnexecution.control;
 
-import org.apache.beam.sdk.fn.function.ThrowingConsumer;
-import org.apache.beam.sdk.util.ThrowingSupplier;
+import javax.annotation.concurrent.ThreadSafe;
 
-/** Control client pool that exposes a source and sink of control clients. */
-public interface ControlClientPool<T extends InstructionRequestHandler> {
+/**
+ * A pool of control clients that brokers incoming SDK harness connections (in the form of {@link
+ * InstructionRequestHandler InstructionRequestHandlers}.
+ *
+ * <p>Incoming instruction handlers usually come from the control plane gRPC service. Typical use:
+ *
+ * <pre>
+ *   // Within owner of the pool, who may or may not own the control plane server as well
+ *   ControlClientPool pool = ...
+ *   FnApiControlClientPoolService service =
+ *       FnApiControlClientPoolService.offeringClientsToSink(pool.getSink(), headerAccessor)
+ *   // Incoming gRPC control connections will now be added to the client pool.
+ *
+ *   // Within code that interacts with the instruction handler. The get call blocks until an
+ *   // incoming client is available:
+ *   ControlClientSource clientSource = ... InstructionRequestHandler
+ *   instructionHandler = clientSource.get("worker-id");
+ * </pre>
+ *
+ * <p>All {@link ControlClientPool} must be thread-safe.
+ */
+@ThreadSafe
+public interface ControlClientPool {
+
+  /** Sink for control clients. */
+  Sink getSink();
 
   /** Source of control clients. */
-  ThrowingSupplier<T> getSource();
+  Source getSource();
 
-  /** Sink for control clients. */
-  ThrowingConsumer<T> getSink();
+  /** A sink for {@link InstructionRequestHandler InstructionRequestHandlers} keyed by worker id. */
+  @FunctionalInterface
+  interface Sink {
+
+    /**
+     * Puts an {@link InstructionRequestHandler} into a client pool. Worker ids must be unique per
+     * pool.
+     */
+    void put(String workerId, InstructionRequestHandler instructionHandler) throws Exception;
+  }
+
+  /** A source of {@link InstructionRequestHandler InstructionRequestHandlers}. */
+  @FunctionalInterface
+  interface Source {
 
+    /**
+     * Retrieves the {@link InstructionRequestHandler} for the given worker id, blocking until
+     * available. Worker ids must be unique per pool. A given worker id must not be requested
```

Comment:
<!--new_thread; commit:d5f16c195fc483161fb1bfb4020edf633708d3b7; resolved:0-->
This gives the impression that if the worker never becomes available, the call never returns. Is this the case?

Also, minor note: this semantics (single get() and put() per workerId) makes me think that perhaps "offer/remove" would be better terminology than "get/put" - or, at least, "remove" would be better than "get".

Reply:
Yes, the call will never return if this worker is never made available or is never explicitly failed. I'll add an explicit comment and rename the method.

I ended up calling this `take`.